design/overview.md

```diff
@@ -73,4 +78,14 @@ Note: X% is some security factor for bootstrapping. 67% is sufficient to release
 
 ## Relaying Cosmos to ETH
 
-**TODO**
\ No newline at end of file
+* Someone (permissionlessly) sends MsgSendToEth this creates an entry in the 'peggy bridge tx pool' this pool is in the Cosmos chain store and contains all transactions not yet bridged. Funds of the specified demon are then burned from the users account if that demon is valid to bridge, otherwise the tx is rejected. 
+* Someone (permissionlessly) sends a MsgRequestBatch, this produces a new `Transaction batch` in the `Peggy Bridge Tx pool`. The creation of this batch occurs in ComsosSDK and is entierly deterministic, ordering transactions from highest to lowest fee and including them in the `Transaction batch` until there are no more transactions or the batch size limit is reached.
+* The `Orchestrator Set` observes the `Peggy Bridge Tx pool` via REST endpoints. When a new batch appears they sign the batch and submit a MsgConfirmBatch containing their Ethereum signature over the batch.
+* Once enough signatures are collected (66% of voting power) any `Orcestrator` can collect these signatures from the REST endpoint, assemble an Ethereum transaction and submit the batch.  
+* The `Orchestrator Set` observes the Ethereum blockchain via their trusted Ethereum node. Once the batch has entered the blockchain and `EthBlockDelay` has elapsed an `Orchestrator` will send MsgBatchInChain. Containing a signed attestation that the block has entered the Ethereum chain, once 66% of all voting power has confirmed the batch the batch is marked as complete. The batch is removed from the `Peggy Bridge Tx pool` and the MsgSendToEth flow is now fully complete.
+* It may be possible for a `TransactionBatch` to be created that is not profitible to submit. This can occur when gas prices are high or the average fee in the bridge is low. In this case the batch will either wait long enough for the gas situation to change or a new MsgRequestBatch will be greated. Exact behavior here will have to take into account a few questions 'can the previous batch be submitted? (check existing signatures in the chain)', 'is the fee to transaction ratio in the new block higher than the old block?' and 'Was the previous batch full or partially empty?'
```

Comment:
Can you move this to a questions/details section below?

It is nice to sketch out the happy path here, then leave the open questions (edge cases) documented. We can dive into those later but they don't matter for the first pass, and keeping them here will be somewhat distracting when we just want to code the happy path

Reply:
Actually all of this. Maybe just make a label "Details and Edge Cases" containing these points

Reply:
As we discussed this morning on the phone, I'm having a hard time understanding why Peggy needs to have all this knowledge about whether a batch is profitable or not. I think we should work backwards from a simple naive implementation before adding all of this logic.

Reply:
How about this:

- Transactions go into the `Peggy Bridge Tx pool`
- The Cosmos state also contains a `LastTxNonce` variable
- Every block, the Cosmos state machine takes the *n* highest fee transactions in the `Peggy Bridge Tx pool` and puts them in a tx batch. The nonces in this batch start from `LastTxNonce`. The orchestrators sign it etc.
- The transactions in the batch are removed from the `Peggy Bridge Tx pool`
- The `LastTxNonce` variable is updated to the highest nonce in the batch
- The batch is added to the `Peggy Bridge Batch pool`

At this point, someone may submit the batch, or they may not. The Cosmos state machine won't know for a while (`EthBlockDelay`). Since the transactions in it have been removed from the `Peggy Bridge Tx pool`, they will not be included in any other batches, thus avoiding a double spend.

Next block, another batch is put together, following the same procedure, starting from the new higher `LastTxNonce`. If this batch is submitted before the previous batch, the previous batch will no longer be submittable because the nonces will be too low. Thus, there's another set of steps in the Cosmos state machine that is done every block:

- We check the `state_lastTxNonce` variable from the contract.
- We check the `Peggy Bridge Batch pool`. Any batches with transactions with nonces under `state_lastTxNonce` are put back in the `Peggy Bridge Tx pool`

Reply:
I'm putting the above scheme in here, not because I think it is the best possible scheme, but because I think it is the simplest possible scheme that works. I feel that we should approach the design process from the angle of trying to find the easiest thing and then optimizing later. Let me know if you agree with that approach. 

If anyone can find a way that the above scheme breaks, OR a way that it could be made even simpler, let me know. Otherwise I would vote for coding the simplest thing and then optimizing it for throughput later (for example allowing multiple batches per block).

Reply:
@jtremback I believe this is about **setting the scope for the MVP** only and that should be happy path. I am sure we need to revisit and reiterate on the unhappy scenarios a couple of times. I am glad and appreciate that you have ideas already. 😄 
Let's track that unhappy scenarios exists. Collect them (maybe in a bigger group), discuss + priotize and then address them with a new spec/ proposal.

Reply:
@jtremback  and I talked about this some.

I think the state machine needs to know about the profitability of batches in order to ensure that an active bridge eventually gets unstuck without anyone eating the gas cost. 

With Jehan's current state machine it is possible to have infinite batches and all of them will remain unprofitable. Especially since we can't cherry pick the most profitable transactions out of previously `Confirmed` batches since they may be submitted unprofitably to double spend.

Reply:
Cool, thanks that's definitely a flaw in my scheme. Just to be clear, the scenario you're referring to is one where many unprofitable txs are being attempted, and only a few profitable ones. Then, every batch will contain enough unprofitable txs so that it is itself not profitable. 

This problem goes away once enough volume of profitable txs has been reached, but that's a pretty high volume and is likely not to be reached a lot of the time.

Your solution seems to deal with this problem, and I now understand why you wrote it the way you did.